nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb

```diff
@@ -10,13 +10,41 @@
 package scala.scalanative
 package unsafe
 
-sealed trait CFuncPtr
+import scala.language.implicitConversions
+import scala.scalanative.annotation.alwaysinline
+import scala.scalanative.runtime.Boxes._
+import scala.scalanative.runtime.{RawPtr, intrinsic}
+
+sealed trait CFuncPtr {
+  def rawptr: RawPtr
+}
+
+object CFuncPtr {
+  @alwaysinline def fromPtr[F <: CFuncPtr](ptr: Ptr[Byte])(
+      implicit tag: Tag.CFuncPtrTag[F]): F =
+    tag.fromRawPtr(ptr.rawptr)
+
+  @alwaysinline def toPtr(ptr: CFuncPtr): Ptr[Byte] = {
+    boxToPtr[Byte](ptr.rawptr)
+  }
+}
 
 % for N in range(0, 23):
-%   args = ", ".join("arg" + str(i) + ": T" + str(i) for i in range(1, N+1))
+%   args 				 = ", ".join("arg" + str(i) + ": T" + str(i) for i in range(1, N+1))
+%   allTps    	 = ", ".join(["T" + str(i) for i in range(1, N+1)] + ["R"])
+%   evidences    = ", ".join(["ev{}: Tag[T{}]".format(i, i) for i in range(1, N+1)] + ["evRet: Tag[R]"])
+%   CFuncPtrN    = "CFuncPtr{}[{}]".format(N, allTps)
+%   FunctionN    = "Function{}[{}]".format(N, allTps)
 
-trait CFuncPtr${N}[${", ".join(["T" + str(i) for i in range(1, N+1)] + ["R"])}] extends CFuncPtr {
-  def apply(${args}): R
+final class ${CFuncPtrN}(val rawptr: RawPtr) extends CFuncPtr {
+  def apply(${args})(implicit ${evidences}): R = intrinsic
 }
 
+object CFuncPtr${N}{
+  implicit def fromScalaFunction[${allTps}](fn: ${FunctionN}): ${CFuncPtrN} = intrinsic
```

Comment:
I wonder whether we should also take `evidence`s of `Tag`s here as well. We might not need them now in the codegen, but I'm not sure that will never be the case. Also, even if we don't use them, it would act as a nice compile-time check that the function's parameter types are static enough (if they're generic type params, the codegen would not do the right thing, I believe, and that would be caught by no Tag being found).

Reply:
Should we also move it to generic `object CFuncPtr`? It would allow doing single import to allow implicit conversion of all method types

Reply:
No, I don't think so. You typically won't need any import anyway, because if the expected type is `CFunctPtr2`, then the `implicit`s in the `object CFunctPtr2` are in scope without import.